### ISSUE TITLE
Directory name is not escaped in redirects

### ISSUE THREAD

**Issue by cees-elzinga** [embedthis/goahead-gpl#7]
The goahead webserver featured last weekend in a hacking competition organised by PPP (https://plaidctf.com/). Competitors were told that the intended solution was to exploit a bug in `file.c`.

I participated on the competition and found two issues. This report contains an issue with unescaped filenames in redirects. I think the implications are limited, but it could be exploited under certain conditions.

## Environment

GoAhead Version: latest (github)
O/S: Linux Ubuntu: 20.04
Hardware: PC

## Problem Description

When a client requests a static directory the server redirects the clients to directory+`websIndex`:

https://github.com/embedthis/goahead-gpl/blob/master/src/file.c#L59

This can be a problem when dealing with 'malicious' directories. An example is a  directory name that contains HTML code:

```
cd webroot
mkdir 'aaaa"><script>'
```

When trying to redirect this page the javascript code is included in the response and gets executed:

```
> GET /pub/aaaa%22%3e%3cscript%3e HTTP/1.1
> Host: localhost:8001
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 302 Redirect
< Content-Length: 238
< Connection: keep-alive
< Location: http://localhost:8001/pub/aaaa"><script>/index.html
< X-Frame-Options: SAMEORIGIN
<html><head></head><body>
        This document has moved to a new <a href="http://localhost:8001/pub/aaaa"><script>/index.html">location</a>.
```

The solution is, probably, to URL escape the directory in the response. This is also what eg Apache does:

```
> GET /aaaa%22%3e%3cscript%3e HTTP/1.1
> Host: localhost
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Location: http://localhost/aaaa%22%3e%3cscript%3e/
< Content-Length: 323
< Content-Type: text/html; charset=iso-8859-1
< 
<p>The document has moved <a href="http://localhost/aaaa%22%3e%3cscript%3e/">here</a>.</p>
```

## Impact / further notes

 - It might also be possible to inject/mess with HTTP headers. I did a quick test with a `\n` but was unable to request the file (404 not found). However, with more effort it might be possible to confuse a client who is parsing these headers.

**Comment by mobsense** [comment]
Thanks for reporting.

However, once an attacker can modify directory names, they can do much more damage that we cannot detect and mitigate. Like create other malicious content, remove content, symlink other directories into the served space.

So while this could be good to escape directories, I don't think it makes much of a dent on the case where the attacker is already inside and has file system permissions.  

Please feel free to push back if you think I'm missing something.

**Comment by cees-elzinga** [comment]
I agree that if someone has filesystem access this issue is moot. Like you said: in that case an attacker can do much more.

I am thinking of a situation where this issue is *combined* with another bug.

For example: a developer creates a website where users can upload files. It has a bug: users can somehow control the directory path where the file is uploaded, and it will create the directory on demand. Depending on other details this might be harmless by itself (eg directory traversals are not possible). But when it's combined with this issue an attacker can now 'upload' javascript by creating malicious directory names.

I believe the same is true for issue #6. I will hijack this comment as it's similar.

For example: a developer creates a websites where users can upload `.txt` files. These files will be served as `text/plain`, so the developer assumes that even if a user uploads (malicious) javascript it will never get executed. But now an attacker uses the race to adjust the content length and inject a fake, second, HTTP response with full control over the HTTP headers and content. Depending on the browser/client it might use this second, injected, HTTP response and execute the javascript.

(This is simplified setup of the "hacking challenge" I referred to. In that case users could upload raw HTML but a strict `Content-Security-Policy` header prevented execution of the javascript. The solution was to use the race to inject a second HTTP response without a `Content-Security-Policy` header. There is a little bit more info here: https://gist.github.com/bluepichu/6898d0f15f9b58ba5a0571213c3896a2).
